### PR TITLE
Handle setTimeout called when tab is shown

### LINKED ISSUE #6227: 0.112 Android app not reconnecting
## Checklist

- [x] I have updated to the latest available Home Assistant version.
- [x] I have cleared the cache of my browser.
- [x] I have tried a different browser to see if it is related to my browser.

## The problem

When the app was in background for a while, it will not reconnect. It says it's reconnecting but will only after force closing the app. 

Might be related to the newly introduced feature which is supposed to freeze the connection after 5 min. 

## Environment

- Home Assistant release with the issue:
- Last working Home Assistant release (if known):
- Browser and browser version:
- Operating system:

## State of relevant entities

```yaml

```

## Problem-relevant configuration

```yaml

```

## Javascript errors shown in your browser console/inspector

```txt

```

### PR DESCRIPTION
**Pull request by balloob**
## Proposed change

On Android Chrome, the tab is frozen after 5 minutes, causing our timeout not to execute until it resumes again.

We can use this API https://github.com/WICG/page-lifecycle to see if we are being frozen and suspend it then.

We put a guard to make sure that if we handle setTimeout action, the tab is actually still hidden.

On resume, we check if we are visible again and unsuspend, we also added a `focus` listener for edge cases. 

Patch with debug statements I used to debug this https://gist.github.com/balloob/3464f55b09442e5f71f72ecf81533ead . Make sure you enable timestamps in Chrome dev console

image

## Type of change

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by balloob** [comment]
I did these tests with the Android app which loads the frontend in a webview. We could have the external bus hook into the Android app lifecycle. Do the same on Android.

However, I am afraid that Chrome on Android will have similar behavior. We should test that.

**Comment by balloob** [comment]
From [StackOverflow](https://stackoverflow.com/a/44035980):

> Visibility of webview is calculated as:
> `visible = !paused && (!attached to window || window visible)`
> In other words, the window visible state is ignored if webview is detached. 

@JBassett do we detach our webview when our app is backgrounded?

**Comment by JBassett** [comment]
@balloob we don't detach the view.  However I was looking over the WebView documentation and found [onPause](https://developer.android.com/reference/android/webkit/WebView#onPause()) and [onResume](https://developer.android.com/reference/android/webkit/WebView#onResume()).  I'll see if I can add those hooks to the WebView and see if that helps.  I'm guessing that's how the paused state is derived.

**Comment by JBassett** [comment]
Doesn't appear to help with the pause/resume.  It wouldn't surprise me if the OS is pausing/killing the app since we aren't doing anything to keep the app running after being closed like a wake lock.

**Comment by larena1** [comment]
https://stackoverflow.com/questions/52028940/how-can-i-make-webview-keep-a-video-or-audio-playing-in-the-background

Could be worth a try

**Comment by balloob** [comment]
So I have one last addition for this fix. It's a way out in case our recover triggers don't trigger. The moment we suspend, we're going to overlay a page that will say that the app is suspended and "tap to continue". That way even if it doesn't recover, there is an alternative.

This is temporary styling, text and it should be translatable. Hoping @bramkragten will pick it up on Monday 😅 

<details><summary>Screenshot</summary>

Screen Shot 2020-06-27 at 13 56 18

</details>

**Comment by JBassett** [comment]
So after more observation I think this is happening on desktop as well.  When tab is sitting idle the connection is disconnected after 5 minutes as well even if the tab is focused.

**Comment by bramkragten** [comment]
> So after more observation I think this is happening on desktop as well. When tab is sitting idle the connection is disconnected after 5 minutes as well even if the tab is focused.
> Chome on Linux

So the tab is visible and focussed but the connection is still disconnected? That would be a bug of Chrome... Let's give ourself some more time to get this right 😄

**Comment by bramkragten** [comment]
For Chrome we can use the `freeze` and `resume` events <https://github.com/WICG/page-lifecycle>.
This is only needed for the connection, if the panel doesn't get disconnected while frozen that doesn't matter.

File: src/layouts/home-assistant.ts [hunk 1]
```diff
@@ -148,31 +158,72 @@ export class HomeAssistantAppEl extends HassElement {
...
+      this._hiddenTimeout = undefined;
     }
+    // Unsuspend the reconnect
+    if (this._visiblePromiseResolve) {
+      this._visiblePromiseResolve();
+      this._visiblePromiseResolve = undefined;
+    }
+  }
+
+  static get styles() {
```
Comment (balloob) [inline src/layouts/home-assistant.ts:211]: We can drop this.

**Comment by balloob** [comment]
Looks good.

**Pull request merged**